File: lambdas/websocket/disconnect.py
import os
import json
import boto3

CONNECTIONS_TABLE = os.getenv('CONNECTIONS_TABLE', 'ConnectionsTable')
WEBSOCKET_ENDPOINT = os.getenv('WEBSOCKET_ENDPOINT', 'YOUR_WEBSOCKET_ENDPOINT')

dynamodb = boto3.resource('dynamodb')
connections_table = dynamodb.Table(CONNECTIONS_TABLE)
apigatewaymanagementapi = boto3.client('apigatewaymanagementapi', endpoint_url=WEBSOCKET_ENDPOINT)

def handler(event, context):
    print(f'Received event: {event}')
    connection_id = event['requestContext']['connectionId']

    # Delete the connection from DynamoDB
    delete_connection(connection_id)

def delete_connection(connection_id):
    try:
        response = connections_table.delete_item(
            Key={
                'connectionId': connection_id
            }
        )
        print(f'Successfully deleted connection {connection_id}')
    except Exception as e:
        print(f'Error deleting connection {connection_id}: {str(e)}')

File: lambdas/src/websocket/disconnect.ts
import { APIGatewayEvent, Context, APIGatewayProxyResult } from 'aws-lambda';
import * as AWS from 'aws-sdk';

// Environment variables
const CONNECTIONS_TABLE = process.env.CONNECTIONS_TABLE || 'ConnectionsTable';
const WEBSOCKET_ENDPOINT = process.env.WEBSOCKET_ENDPOINT || 'YOUR_WEBSOCKET_ENDPOINT';

// Initialize DynamoDB DocumentClient and API Gateway Management API
const dynamodb = new AWS.DynamoDB.DocumentClient();
const connectionsTable = CONNECTIONS_TABLE;
const apigatewaymanagementapi = new AWS.ApiGatewayManagementApi({ endpoint: WEBSOCKET_ENDPOINT });

export const handler = async (event: APIGatewayEvent, context: Context): Promise<APIGatewayProxyResult> => {
    console.log(`Received event: ${JSON.stringify(event)}`);
    const connectionId = event.requestContext.connectionId;

    // Delete the connection from DynamoDB
    await deleteConnection(connectionId as string);

    return {
        statusCode: 200,
        body: 'Disconnected successfully.',
    };
};

const deleteConnection = async (connectionId: string): Promise<void> => {
    try {
        await dynamodb.delete({
            TableName: connectionsTable,
            Key: {
                connectionId,
            },
        }).promise();
        console.log(`Successfully deleted connection ${connectionId}`);
    } catch (error) {
        console.error(`Error deleting connection ${connectionId}:`, error);
    }
};
